cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -24,16 +20,148 @@ use std::collections::HashSet;
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub async fn matching_create_contract<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<(Transaction, Address)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    let (transaction, receipt) =
+        matching_transaction_and_receipt(blockchain_connector, start_of_swap, |transaction| {
+            // transaction.to address is None if, and only if, the transaction
+            // creates a contract.
+            transaction.to.is_none() && transaction.input == bytecode
+        })
+        .await?;
+
+    match receipt.contract_address {
+        Some(location) => Ok((transaction, location)),
+        None => Err(anyhow::anyhow!("contract address missing from receipt")),
+    }
+}
+
+pub async fn matching_event<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    event: Event,
+) -> anyhow::Result<(Transaction, Log)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction_and_log(
+        blockchain_connector.clone(),
+        start_of_swap,
+        event.topics.clone(),
+        |receipt| find_log_for_event_in_receipt(&event, receipt),
+    )
+    .await
+}
+
+/// Fetch receipt from connector using transaction hash.
+async fn fetch_receipt<C>(blockchain_connector: C, hash: Hash) -> anyhow::Result<TransactionReceipt>
+where
+    C: ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>,
+{
+    let receipt = blockchain_connector
+        .receipt_by_hash(hash)
+        .compat()
+        .await?
+        .ok_or_else(|| {
+            anyhow::anyhow!(
+                "Could not get transaction receipt for transaction {:x}",
+                hash
+            )
+        })?;
+    Ok(receipt)
+}
+
+fn find_log_for_event_in_receipt(event: &Event, receipt: TransactionReceipt) -> Option<Log> {
+    match event {
+        Event { topics, .. } if topics.is_empty() => None,
+        Event { address, topics } => receipt.logs.into_iter().find(|log| {
+            if address != &log.address {
+                return false;
+            }
+
+            if log.topics.len() != topics.len() {
+                return false;
+            }
+
+            log.topics.iter().enumerate().all(|(index, tx_topic)| {
+                let topic = &topics[index];
+                topic.as_ref().map_or(true, |topic| tx_topic == &topic.0)
+            })
+        }),
+    }
+}
+
+pub async fn matching_transaction_and_receipt<C, F>(
```

Comment:
:question: Is this public just so that we can test it?

Reply:
Yes, funny you should ask that ...